### ISSUE TITLE
is the data(ctd) temperature scale correct? (did it get converted twice?)

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1293]
The processing-log states
```
    - 2017-07-16 17:00:13 UTC: `oce.edit(x = ctd, item = "temperature", value = T90fromT68(ctd[["temperature"]]),     reason = "File uses old IPTS-68 convention", person = "Dan Kelley")`
```
but in the source file I see
```
$ grep temperature /Users/kelley/git/oce/create_data/ctd/ctd.cnv 
# name 4 = t068: temperature, IPTS-68 [deg C]
# sensor 0 = Frequency 0  temperature, 1140, 13 Mar 03
```
so shouldn't `oce` have detected and converted the unit by itself? Are we asking users to call `T90fromT68()` themselves? I thought we had decided to handle that internally. I don't have time to look at this today, so this is a to-do for the weekend, probably.

PS. I ran into this because of classes! It was a roundabout thing. Another prof told me of a source that says that sigma-t is 1000*(rho/rho0-1), where rho0 is not 1000kg/m^3, but rather the maximum density at S=0 and std atm pressure, which is not exactly 1000kg/m^3. But Sverdrup, Johnson and Fleming use 1000kg/m^3 (in cgs), after stating that this is the exact value of that max density. If I were defining it, I'd use max density, but I really don't think that is what people do in practice. Convention is confusing (*Plus ca change*)!

**Comment by richardsc** [comment]
I feel there is probably an issue discussion that goes over this, but my memory says that we read and store what is in the file (e.g. IPTS-68) and *don't* convert it by default. Otherwise it gets too confusing.

Of course, the issue then is if you use the object (with the IPTS-68 scale) to calculate any `sw*()` properties (or `gsw`), I think that they will be wrong, since we made sure that all the temperature inputs to the `sw` functions go through a `T68fromT90()`.

That says that we should be checking units for *all* the `sw*()` functions, and converting only if required. But that's probably a mess of coding ...

Maybe we should just convert T68 to T90 when we read the file? Wait, didn't I just argue *against* that? I'm confused ...

**Comment by richardsc** [comment]
A trip down memory lane: Issue #945 "Should read.ctd() promote a calculated ITS-90 temperature to be "the" temperature? "

**Comment by dankelley** [comment]
Note: this data file has
```
# name 4 = t068: temperature, IPTS-68 [deg C]
```
and we presently match for `t068C`, so this is being caught not by a direct (specific) match but rather by an indirect (wider regexp) match, which I would prefer not to be the case. I will add a line to match this directly, and also a line to the documentation to note that it is being matched directly.

It would be great for CR and DK to talk about this in person. I see commented-out code at the end of `R/ctd.sbe.R` that indicates that we used to do a conversion. I know we have gone back and forth on this a lot! But I think things are OK in `read.ctd.sbe` because we now use scales. For example, in `R/ctd.R`, the accessor has code (near line 610 +- 10 or so) as follows:
```R
              } else if (i == "temperature") {
                  scale <- x@metadata$units[["temperature"]]$scale
                  if (!is.null(scale) && "IPTS-68" == scale)
                      T90fromT68(x@data$temperature)
                  else x@data$temperature
 ```
and that means that if users do the "right thing" and use accessors (instead of using `object@data$temperature` then all should be okay.

From what I just wrote, I conclude that `data(ctd)` is, indeed, being computed incorrectly. Well, or it will be, when I ensure that this SBE code `"t068" is recognized, which it will be in a few minutes.

**Comment by richardsc** [comment]
I don't follow your last line of reasoning. If we read in the temperature as it is recorded in the file, and record the correct units and scales to the metadata, then there is no "error" per se. The issue of course is that an object with an IPTS-68 scale should not be passed through other downstream functions (like the `sw*()`, at least without the proper checks in the downstream functions for temperature scale (which I think we're *not* doing right now), or without first calculating the proper ITS-90 temperature (as is done in the creation of the `data(ctd)` object.

The previous discussion indicated that we should at least be spitting a warning that a non ITS-90 temperature was read, which we're not doing now.

**Comment by dankelley** [comment]
The error was that in `create_data/ctd` I was changing the numbers stored in the object.

I agree, we ought to warn. There's the question of when to do it. It could be just when we read in the numbers. But someone might use pre-loaded data (lots of times I share .rda files) so maybe a message should be printed when the `[["temperature"]]` method is used. Which do you think is best?

**Comment by richardsc** [comment]
Ah, I see.

I think the warning should be only when the data are read. Coding warnings into the accessor seems like such an edge case that it will just end up being cruft code where we try and predict future (bad) behaviour with sharing pre-read files.

**Comment by dankelley** [comment]
Thanks. Any advice on what it should say? Right now I have as follows:
```R
Warning message:
In read.ctd.sbe(file, processingLog = processingLog, ...) :
  this CNV file has temperature in the IPTS-68 scale, and this is stored in the object; note that [["temperature"]] and the sw* functions will convert to the modern ITS-90 value
```
It's quite long, but I think I'd appreciate knowing all of that, as a user. Please suggest any edit you like. Also, note that the doc now has as below (again, edits in a comment please)
```R
#' @section A note on scales:
#' The user may encounter data files with a variety of scales for temperature and
#' salinity. Oce keeps track of these scales in the units it sets up for the stored
#' variables. For example, if \code{A} is a CTD object, then
#' \code{A[["temperatureUnit"]]$scale} is a character string that will indicate the scale.
#' Modern-day data will have \code{"ITS-90"} for that scale, and old data may have
#' \code{"IPTS-68"}. The point of saving the scale in this way is so that the various
#' formulas that deal with water properties can account for the scale, e.g. converting
#' from numerical values saved on the \code{"IPTS-68"} scale to the newer scale, using
#' \code{\link{T90fromT68}} before doing calculations that are expressed in
#' terms of the \code{"ITS-90"} scale. This is taken care of by retrieving temperatures
#' with the accessor function, e.g. writing \code{A[["temperature"]]} will either
#' retrieve the stored values (if the scale is ITS-90) or converted values (if
#' the scale is IPTS-68). Even though this procedure should work, users who
#' really care about the details of their data are well-advised to do a couple
#' of tests after examining the first data line of their data file in an editor.
#'
#' Note that reading a file that contains IPTS-68 temperatures produces a warning.
```

PS. I'm asking for edits as a comment here because none of this is committed. Going offline for an hour or so.

**Comment by dankelley** [comment]
I've coded a solution in "develop" commit 2c92f08538a4346bc40b1f386ee5263c3c60785e. Part of it is to rebuild the `ctd` and `ctdRaw` datasets, but it also involved
1. `read.ctd.sbe` now warns if the temperature scale is IPTS-68
2. `tests/testthat/test_accessors.R` accounts for adjusted `data(ctd)` and gets more tests
3. `tests/testthat/test_ctd.R` accounts for adjusted `data(ctd)` and gets more tests

Since I've fixed the issue as stated, I'll go ahead and close it. The wording of the docs and the new warning can be tweaked, of course, and I'd appreciate it if @richardsc would take a look at these (I suggest grepping in the code for words in the previous comment, in case I've made updates that I forgot to note above).
